### PR TITLE
Update to xtra upstream 

### PR DESCRIPTION
**Pull request by Restioson**
Continuation of #2306, but rebased for cleaner history.

### CODE REVIEW

**Review by klochowicz — COMMENTED** [review]
would you mind adding a bit more info in `Update xtra to upstream`, ie. list notable changes in xtra into which we're adapting?

File: Cargo.toml
Comment (klochowicz) [inline Cargo.toml]: are you planning to merge this xtra branch to `master` before merging this PR? I am ok with pinning rev inside the `master` branch, but not that keen on depending on some side branch.

(if yes, please add a TODO about that)

File: daemon/Cargo.toml
Comment (klochowicz) [inline daemon/Cargo.toml:53]: I guess it's really important to mention in the PR description and in the commit message that:
- we're removing xtra metrics (and that the new tracing instrumentation framework will be ready before the new release)
- timeouts are supported inside `xtra` out-of-the-box (new implementation)
- `tracing` got renamed to `instrumentation` 

otherwise, one might think we're losing a lot of features :)

File: daemon/src/lib.rs
Comment (klochowicz) [inline daemon/src/lib.rs]: nit: we typically qualify this one at the callsite, just like other structs from this module

File: daemon/src/projection.rs
Comment (klochowicz) [inline daemon/src/projection.rs:79]: as a user: this is a bit more verbose, but I actually don't mind it as I see what's being returned from that handler!

**Comment by Restioson** [comment]
Sure!

File: xtra-libp2p/src/dialer.rs
Comment (klochowicz) [inline xtra-libp2p/src/dialer.rs]: did you mean to change this line? it probably doesn't belong to this changeset :)

File: xtra-libp2p-offer/Cargo.toml
Comment (klochowicz) [inline xtra-libp2p-offer/Cargo.toml:17]: no need to add "instrumentation" to this one?

File: xtras/src/handler_timeout.rs
Comment (klochowicz) [inline xtras/src/handler_timeout.rs:9]: would you mind adding a bit of docs to this trait and/or method?

File: maker/src/connection.rs
Comment (klochowicz) [inline maker/src/connection.rs]: Please add a comment describing the reason for wrapping this type

**Review by klochowicz — COMMENTED** [review]
looks pretty good, I've added some comments, mostly to aid in readability :)

File: daemon/src/projection.rs
Comment (Restioson) [reply daemon/src/projection.rs:79]: It is a bit more verbose in these kind of places, yea. That is the downside. For historians, the PR that removed the Message trait is https://github.com/Restioson/xtra/pull/62

File: xtra-libp2p/src/dialer.rs
Comment (Restioson) [reply xtra-libp2p/src/dialer.rs]: Thanks, this was by accident

File: xtra-libp2p-offer/Cargo.toml
Comment (Restioson) [reply xtra-libp2p-offer/Cargo.toml:17]: `instrumentation` doesn't affect the public API, so only binaries really care about it being enabled I think.

File: maker/src/connection.rs
Comment (Restioson) [reply maker/src/connection.rs]: Ah, I remember that this is something that we discussed - the alternative to having this type and having its own handler would be to replace the sink somewhere else with just an address send which intentionally swallows the error. The reason for this now is that MessageSink only exists for `Handler<Return = ()>`, to avoid swallowing errors or return values accidentally.

Comment (klochowicz) [reply maker/src/connection.rs]: if this is preserving the old behaviour, and just forcing us to be explicit about dropping the errors, this is also fine!

**Comment by klochowicz** [comment]
I'm afraid you might need to rebase on master (there's lots of conflicts, probably after I merged a huge dependency upgrade recently)....

**Comment by klochowicz** [comment]
as for the supervisor refactor - I recommend leaving that for a separate PR :) then supervisor and adding instrumentation could be looked at independently

**Comment by Restioson** [comment]
> I'm afraid you might need to rebase on master (there's lots of conflicts, probably after I merged a huge dependency upgrade recently)....

Will do!

> as for the supervisor refactor - I recommend leaving that for a separate PR :) then supervisor and adding instrumentation could be looked at independently

I've pushed it to the `supervisor-refactor` on my fork for now, so it's there for when the time comes

File: Cargo.toml
Comment (Restioson) [reply Cargo.toml]: This could get merged https://github.com/Restioson/xtra/pull/112 does, if I remove instrumentation from the timeout handler - the impact of this would be that the "Handler execution timed out" log wouldn't be attached to the handler span. Otherwise, it should wait for that to be merged and I can add a TODO. Maybe this is a small price to pay for getting this merged earlier though? We can always change it back to properly attach the log to the span when #112 does get merged - it should only be a few lines change. Let me know which is best!

File: maker/src/connection.rs
Comment (Restioson) [reply maker/src/connection.rs]: The other change with sinks was about backpressure - now, it will wait for the messages it sends in to be handed before it sends the next.

Comment (Restioson) [reply maker/src/connection.rs]: I'll reimplement this with `sink::unfold` to be equivalent to the old version - throwing away errors and no backpressure. It should be pretty easy to change to logging errors or stopping the sink on errors with an `unfold`-based sink

**Comment by Restioson** [comment]
Could someone enable CI workflows please? :slightly_smiling_face:

File: daemon/Cargo.toml
Comment (bonomat) [reply daemon/Cargo.toml:53]: Can you update changelog.md as well please?

Comment (Restioson) [reply daemon/Cargo.toml:53]: Will do!

Comment (Restioson) [reply daemon/Cargo.toml:53]: Should I mention just the fact that xtra was updated, or also the internal work such as replacing handler timeouts with a solution in `xtras`?

Comment (Restioson) [reply daemon/Cargo.toml:53]: Pushed a minimal update, let me know if I should add more detail to it! My plan is probably to squash the commit into "Update xtra to upstream" via interactive rebase before this is merged.

File: CHANGELOG.md
Comment (bonomat) [inline CHANGELOG.md:12]: Maybe add that with this patch the metrics from xtra are being removed.

**Comment by bonomat** [comment]
> Could someone enable CI workflows please? 🙂

I've added you as a contributor. Hopefully this will help.

**Comment by Restioson** [comment]
Thank you!

File: Cargo.toml
Comment (klochowicz) [reply Cargo.toml]: I would prefer merging this earlier so that it can run on our testnet over the weekend - as you correctly noticed, that span addition can always be added later :)

File: Cargo.toml
Comment (klochowicz) [inline Cargo.toml]: as talked earlier, this not being from `master` is the only blocker for merging this PR.

File: maker/src/contract_setup.rs
Comment (klochowicz) [inline maker/src/contract_setup.rs:106]: if we can easily log the error here, I'd opt for that (as it is an improvement :) ). Perhaps something like:

```suggestion
                    if let Err(e) = taker.send(msg).split_receiver().await {
                        tracing::error!("failed to send contract_setup message to the taker: {e:#});
                    } 
```

(in all instances where you introduced unfolding)

File: maker/src/rollover.rs
Comment (klochowicz) [inline maker/src/rollover.rs:231]: same suggestion as above, instead of dropping the error, we can at least log it.

File: xtras/src/handler_timeout.rs
Comment (klochowicz) [inline xtras/src/handler_timeout.rs:87]: nice to see this covered by a test 😀

**Review by klochowicz — COMMENTED** [review]
Thanks for updates, I really like usage of `unfold` instead of a separate message channel! As stated above, the span for handler timeout can be done as a follow-up when `xtra` has these changes in master.

File: maker/src/contract_setup.rs
Comment (Restioson) [reply maker/src/contract_setup.rs:106]: Can do!

File: xtras/src/handler_timeout.rs
Comment (Restioson) [reply xtras/src/handler_timeout.rs:87]: Haha, I felt it was needed after that!

File: maker/src/contract_setup.rs
Comment (Restioson) [reply maker/src/contract_setup.rs:106]: Ah, we cannot actually do this currently without https://github.com/Restioson/xtra/pull/105, unless we first wait for the message to be handled and then print the result. This has a different backpressure behaviour. On the other hand, it could still log an error if the taker is disconnected right after sending the message, or right before.

Comment (Restioson) [reply maker/src/contract_setup.rs:106]: Additionally, should the sink stop accepting messages if an error occurs? I think this can be done by returning `Err` from `unfold`

Comment (klochowicz) [reply maker/src/contract_setup.rs:106]: interesting idea, although I would probably not go that far yet, as this will indeed change the behaviour. I would like to first get some stats on how often we see errors here.

Comment (klochowicz) [reply maker/src/contract_setup.rs:106]: > Ah, we cannot actually do this currently without [Restioson/xtra#105](https://github.com/Restioson/xtra/pull/105), unless we first wait for the message to be handled and then print the result. This has a different backpressure behaviour. On the other hand, it could still log an error if the taker is disconnected right after sending the message, or right before.

Ok, no worries! we can also print the messages later. happy for it to go without printing these logs :)

**Review by klochowicz — APPROVED** [review]
looks good!
please squash the small fixup commits in the first one, and it's ready to merge :)

**Comment by klochowicz** [comment]
bros r+

**Comment by klochowicz** [comment]
@Restioson could you please rebase on top of latest master and force-push? (I tried to do a merge in github, which fixed the error, but formatting failed). 

Amending the merge commit would work too (by adding the newline).

**Comment by klochowicz** [comment]
@Restioson could you please rebase on top of latest master and force-push? (I tried to do a merge in github, which fixed the error, but formatting failed). 

Amending the merge commit would work too (by adding the newline).

**Pull request merged**
